composer/loggers/wandb_logger.py

```diff
@@ -67,8 +61,19 @@ def epoch_end(self, state: State, logger: Logger) -> None:
         if self._log_artifacts:
             self._upload_artifacts()
 
+    def training_end(self, state: State, logger: Logger) -> None:
+        del state, logger  # unused
+        if self._log_artifacts:
+            self._upload_artifacts()
+
     def _upload_artifacts(self):
         # Scan the run directory and upload artifacts to wandb
+        # On resnet50, _log_artifacts() caused a 22% throughput degradation
```

Comment:
:(

Any solutions here? Would prefer for something non-critical (e.g. logging artifacts) inadvertently cause a slowdown for users.

Reply:
Yup! `_upload_artifacts` is called only once every 100 (or a user configurable amount) batches, so this cost is amortized and becomes negligable.